File: node/miner.go
package node

import (
	"fmt"
	"math/big"
	"os"
	"time"

	"zephyria/core"
	"zephyria/state"
	"zephyria/types"
	ztypes "zephyria/types"

	"github.com/ethereum/go-ethereum/common"
	ethtypes "github.com/ethereum/go-ethereum/core/types"
	"github.com/ethereum/go-ethereum/crypto"
	"github.com/syndtr/goleveldb/leveldb"
)

// OptResult holds the result of an optimistic (speculative) block execution.
type OptResult struct {
	ParentHash common.Hash
	Txs        []*ethtypes.Transaction
	Receipts   []*types.Receipt
	Root       common.Hash
	GasUsed    uint64
	OptState   *state.StateDB
}

// StartMining starts the block production loop and PoH metronome.
func (n *Node) StartMining() {
	fmt.Println("Miner Started")

	// Determine starting VDF seed and slot from tip
	head := n.bc.CurrentBlock()
	seed := head.Hash().Bytes()
	slot := head.Header.Number.Uint64()

	// If head has VDF output, use that as seed for continuity
	vdfSize := (n.engine.VDFIterations / n.engine.VDFCheckpointInterval) * 32
	if len(head.Header.ExtraData) >= vdfSize {
		seed = head.Header.ExtraData[vdfSize-32 : vdfSize]
	}

	n.engine.Metronome.Start(seed, slot)

	n.wg.Add(1)
	go func() {
		defer n.wg.Done()
		n.loop(n.txCh, n.executor)
	}()
}

func (n *Node) loop(txCh chan *ethtypes.Transaction, executor *core.Executor) {
	batch := make([]*ethtypes.Transaction, 0)

	round := 0
	for {
		select {
		case <-n.stopCh:
			return

		// Wake up signals
		case <-n.txCh:
			// Legacy internal signal (Stake/Unstake)
			// Drain logic
		loopDrain:
			for {
				select {
				case <-n.txCh:
				default:
					break loopDrain
				}
			}
			// Fallthrough to process pool

		case <-n.poolCh:
			// TxPool signal (RPC)
			// Drain
		poolDrain:
			for {
				select {
				case <-n.poolCh:
				default:
					break poolDrain
				}
			}

			// Process Pending Transactions
			batch = n.txPool.Pending()
			if len(batch) > 0 {
				fmt.Fprintf(os.Stderr, "[Miner] Picked up %d txs from pool\n", len(batch))
			}

			if len(batch) > 0 {
				// State Pre-fetching: Warm up cache for all txs in batch
				go func(txs []*ethtypes.Transaction) {
					addrs := make([]common.Address, 0, len(txs)*2)
					for _, tx := range txs {
						if tx.To() != nil {
							addrs = append(addrs, *tx.To())
						}
					}
					n.state.Prefetch(addrs)
				}(batch)

				if len(batch) > 0 {
					go n.triggerOptimisticExecution(batch)
				}
			}

		case slot := <-n.engine.Metronome.TickCh:
			// Dynamic Validator Sync
			// We need to ensure we have the latest set before checking leader
			select {
			case <-n.stopCh:
				return
			default:
			}

			n.stateLock.Lock()
			if err := n.engine.SyncValidators(n.state); err != nil {
				fmt.Printf("Validator Sync Error: %v\n", err)
			}
			n.stateLock.Unlock()

			// Check if it's our turn to propose
			parent := n.bc.CurrentBlock()
			nextBlockNum := new(big.Int).Add(parent.Header.Number, common.Big1)

			proposer := n.engine.GetLeader(slot, parent.Hash())
			myAddr := crypto.PubkeyToAddress(n.engine.PrivateKey().PublicKey)

			if proposer == myAddr {
				miningStart := time.Now()
				// We are the Proposer for this slot!
				if len(batch) > 0 {
					fmt.Fprintf(os.Stderr, "[Miner] Proposing for Slot #%d (My Turn! Batch: %d)\n", slot, len(batch))
				}
				// Re-filter batch against LATEST state before proposing
				// This removes txs that were already included by peers while we waited
				select {
				case <-n.stopCh:
					return
				default:
				}
				n.stateLock.Lock()
				signer := ethtypes.LatestSigner(n.netCfg.ChainConfig())
				validTxs := make([]*ethtypes.Transaction, 0)
				remainingTxs := make([]*ethtypes.Transaction, 0)
				pendingNonces := make(map[common.Address]uint64)

				limit := 2800
				for _, tx := range batch {
					sender, err := ethtypes.Sender(signer, tx)
					if err != nil {
						continue
					}

					nonce, ok := pendingNonces[sender]
					if !ok {
						nonce = n.state.GetNonce(sender)
					}

					if tx.Nonce() == nonce && len(validTxs) < limit {
						validTxs = append(validTxs, tx)
						pendingNonces[sender] = nonce + 1
					} else if tx.Nonce() > nonce {
						// Future tx, keep in batch
						remainingTxs = append(remainingTxs, tx)
					} else {
						// tx.Nonce() < nonce, already used by peer, DROP
					}
				}
				batch = remainingTxs
				// Keep stateLock LOCKED for ApplyBlock!

				if len(validTxs) > 0 {
					toProcess := validTxs
					start := time.Now()

					// OPTIMISTIC FAST PATH
					n.optMu.Lock()
					opt := n.optResult
					n.optResult = nil // consume it
					n.optMu.Unlock()

					var blockReceipts []*types.Receipt
					var blockRoot common.Hash
					var gasUsed uint64

					// Create a header for the current block proposal
					header := &ztypes.Header{
						ParentHash: parent.Hash(),
						Number:     nextBlockNum,
						Time:       uint64(time.Now().Unix()),
						Coinbase:   n.netCfg.Coinbase,
						GasLimit:   n.netCfg.Params.DefaultGasLimit,
						BaseFee:    core.CalcBaseFee(n.netCfg.ChainConfig(), parent.Header),
					}

					// Compute VDF
					header.ExtraData = n.engine.ComputeVDF(parent)

					if opt != nil && opt.ParentHash == parent.Hash() && n.compareTxs(opt.Txs, toProcess) {
						fmt.Printf("\033[1;33m[⚡] Optimistic Cache Hit\033[0m (Block #%d)\n", nextBlockNum)
						opt.OptState.Merge()
						blockReceipts = opt.Receipts
						// Recalculate root from base state because OptState (overlay) cannot compute new root
						blockRoot = n.state.IntermediateRoot(false)
						gasUsed = opt.GasUsed
					} else {
						receipts, root, err := n.executor.ApplyBlock(n.state, header, toProcess)
						if err != nil {
							fmt.Println("Block execution failed:", err)
							return
						}
						blockReceipts = receipts
						blockRoot = root
						gasUsed = getGasUsed(receipts)
					}

					header.VerkleRoot = blockRoot
					header.GasUsed = gasUsed
					n.lastExecTime = time.Since(start)

					// 5. Seal (Consensus)
					block := ztypes.NewBlock(header, toProcess)
					if err := n.engine.Seal(block, slot); err != nil {
						fmt.Println("Failed to seal block:", err)
						return
					}

					// PIPELINING: 6. Broadcast IMMEDIATELY after sealing
					n.p2p.BroadcastBlock(block)

					// VOTOR: Vote for own block
					if vote, err := n.engine.CreateVote(block.Hash(), block.Header.Number.Uint64()); err == nil {
						n.votePool.AddVote(vote) // Add own vote
						// Check QC immediately (Single validator case)
						if reached, _, bitmask := n.votePool.CheckQuorum(block.Hash()); reached {
							fmt.Printf("\033[1;35m[⚡] BLOCK FINALIZED\033[0m: %s | QC Reached via %x (Self)\n", block.Hash().Hex()[:10], bitmask)
						}
						n.p2p.BroadcastVote(vote)
					}
					// 7. Persistence (Synchronous for safety)
					batchDB := new(leveldb.Batch)
					if _, err := n.state.Commit(n.bc.Database(), batchDB, block.Header.Number.Uint64()); err != nil {
						fmt.Println("State Commit failed:", err)
					}
					if err := n.bc.Database().Write(batchDB, nil); err != nil {
						fmt.Println("State persistence failed:", err)
					}
					if err := n.bc.AddBlock(block, blockReceipts); err != nil {
						fmt.Println("Failed to add block:", err)
					}

					// SYNC POH METRONOME
					vdfSize := (n.engine.VDFIterations / n.engine.VDFCheckpointInterval) * 32
					if len(header.ExtraData) >= vdfSize {
						lastVDF := header.ExtraData[vdfSize-32 : vdfSize]
						n.engine.Metronome.Sync(lastVDF, header.Number.Uint64())
					}

					// Update Metrics
					DefaultMetrics.IncBlocks()
					DefaultMetrics.IncTxs(len(toProcess))

					// Log
					elapsed := time.Since(miningStart)
					fmt.Fprintf(os.Stderr, "\033[1;32m[✓] Block Mined\033[0m #%d | \033[1;36mHash:\033[0m %s | \033[1;36mTxs:\033[0m %d | \033[1;36mGas:\033[0m %d | \033[1;35mTime:\033[0m %v (Pipelined)\n",
						block.Header.Number, block.Hash().Hex()[:10], len(toProcess), header.GasUsed, elapsed)

					// Update TxPool
					n.txPool.StateUpdate(func() *state.StateDB { return n.state })
				} else {
					// Proposer but no txs? Produce empty block to keep chain moving
					n.optMu.Lock()
					n.optResult = nil // clear cache
					n.optMu.Unlock()

					header := &ztypes.Header{
						ParentHash: parent.Hash(),
						Number:     nextBlockNum,
						Time:       uint64(time.Now().Unix()),
						Coinbase:   n.netCfg.Coinbase,
						GasLimit:   n.netCfg.Params.DefaultGasLimit,
						BaseFee:    core.CalcBaseFee(n.netCfg.ChainConfig(), parent.Header),
					}
					header.ExtraData = n.engine.ComputeVDF(parent)

					receipts, root, err := n.executor.ApplyBlock(n.state, header, []*ethtypes.Transaction{})
					if err == nil {
						header.VerkleRoot = root
						header.GasUsed = getGasUsed(receipts)
						block := ztypes.NewBlock(header, []*ethtypes.Transaction{})
						n.engine.Seal(block, uint64(round))
						n.p2p.BroadcastBlock(block)

						// 7. Persistence (Synchronous)
						batchDB := new(leveldb.Batch)
						n.state.Commit(n.bc.Database(), batchDB, block.Header.Number.Uint64())
						n.bc.Database().Write(batchDB, nil)
						n.bc.AddBlock(block, receipts)

						// SYNC POH METRONOME
						vdfSize := (n.engine.VDFIterations / n.engine.VDFCheckpointInterval) * 32
						if len(header.ExtraData) >= vdfSize {
							lastVDF := header.ExtraData[vdfSize-32 : vdfSize]
							n.engine.Metronome.Sync(lastVDF, header.Number.Uint64())
						}

						// Update Metrics
						DefaultMetrics.IncBlocks()

						// Log
						elapsed := time.Since(miningStart)
						fmt.Fprintf(os.Stderr, "\033[1;32m[✓] Block Mined\033[0m #%d | \033[1;36mHash:\033[0m %s | \033[1;36mTxs:\033[0m 0 | \033[1;36mGas:\033[0m %d | \033[1;35mTime:\033[0m %v (Empty)\n",
							block.Header.Number, block.Hash().Hex()[:10], header.GasUsed, elapsed)
					}
				}
				n.stateLock.Unlock() // Balanced unlock
			} else {
				// We are NOT the proposer. Just wait.
			}
		}
	}
}

func (n *Node) triggerOptimisticExecution(txs []*ethtypes.Transaction) {
	if len(txs) == 0 {
		return
	}

	parent := n.bc.CurrentBlock()
	nextNum := new(big.Int).Add(parent.Header.Number, common.Big1)

	// Create a mock header for optimistic execution
	header := &ztypes.Header{
		ParentHash: parent.Hash(),
		Number:     nextNum,
		Time:       uint64(time.Now().Unix()),
		Coinbase:   n.netCfg.Coinbase,
		GasLimit:   n.netCfg.Params.DefaultGasLimit,
		BaseFee:    core.CalcBaseFee(n.netCfg.ChainConfig(), parent.Header),
	}

	// Re-filter batch for speculative execution
	n.stateLock.Lock()
	signer := ethtypes.LatestSigner(n.netCfg.ChainConfig())
	validTxs := make([]*ethtypes.Transaction, 0)
	pendingNonces := make(map[common.Address]uint64)
	for _, tx := range txs {
		sender, err := ethtypes.Sender(signer, tx)
		if err != nil {
			continue
		}
		nonce, ok := pendingNonces[sender]
		if !ok {
			nonce = n.state.GetNonce(sender)
		}
		if tx.Nonce() == nonce {
			validTxs = append(validTxs, tx)
			pendingNonces[sender] = nonce + 1
		}
	}
	n.stateLock.Unlock()

	if len(validTxs) == 0 {
		return
	}

	overlay := n.state.NewOverlay()
	receipts, root, err := n.executor.ApplyBlock(overlay, header, validTxs)
	if err != nil {
		return
	}

	n.optMu.Lock()
	defer n.optMu.Unlock()
	n.optResult = &OptResult{
		ParentHash: parent.Hash(),
		Txs:        validTxs,
		Receipts:   receipts,
		Root:       root,
		GasUsed:    getGasUsed(receipts),
		OptState:   overlay,
	}
}

func (n *Node) compareTxs(a, b []*ethtypes.Transaction) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i].Hash() != b[i].Hash() {
			return false
		}
	}
	return true
}

func getGasUsed(receipts []*ztypes.Receipt) uint64 {
	if len(receipts) == 0 {
		return 0
	}
	return receipts[len(receipts)-1].CumulativeGasUsed
}

File: src/node/miner.zig
const std = @import("std");
const types = @import("core").types;
const consensus = @import("consensus");
const zelius = @import("consensus").zelius;
const vrf = @import("consensus").vrf;
const vdf = @import("consensus").vdf;
const blockchain = @import("core").blockchain;
const tx_pool = @import("core").tx_pool;
const executor = @import("core").executor;
const state = @import("core").state;
const p2p = @import("p2p");

pub const Miner = struct {
    allocator: std.mem.Allocator,
    chain: *blockchain.Blockchain,
    tx_pool: *tx_pool.TxPool,
    engine: *zelius.ZeliusEngine,
    executor: *executor.Executor,
    state: *state.State,

    running: *std.atomic.Value(bool),
    validator_addr: types.Address,
    p2p_server: ?*p2p.Server,

    pub fn init(
        allocator: std.mem.Allocator,
        chain: *blockchain.Blockchain,
        pool: *tx_pool.TxPool,
        engine: *zelius.ZeliusEngine,
        exec: *executor.Executor,
        state_obj: *state.State,
        addr: types.Address,
        running_flag: *std.atomic.Value(bool),
    ) !*Miner {
        const self = try allocator.create(Miner);
        self.* = Miner{
            .allocator = allocator,
            .chain = chain,
            .tx_pool = pool,
            .engine = engine,
            .executor = exec,
            .running = running_flag,
            .validator_addr = addr,
            .state = state_obj,
            .p2p_server = null,
        };
        return self;
    }

    pub fn deinit(self: *Miner) void {
        self.allocator.destroy(self);
    }

    pub fn set_p2p(self: *Miner, server: *p2p.Server) void {
        self.p2p_server = server;
    }

    pub fn start(self: *Miner) !void {
        std.debug.print("[Miner] Starting block production loop for {f}\n", .{self.validator_addr});

        while (self.running.load(.seq_cst)) {
            // 1. Get current head
            const parent = self.chain.current_block orelse return error.NoGenesis;
            const next_number = parent.header.number + 1;

            // 2. Check VRF Eligibility
            const eligible = try self.check_eligibility(parent);
            if (!eligible) {
                // Wait for next slot
                std.Thread.sleep(1000 * std.time.ns_per_ms);
                continue;
            }

            std.debug.print("[Miner] Eligible for block {d}, starting production...\n", .{next_number});

            // 3. Produce Block
            const block = try self.produce_block(parent);

            // 4. Seal Block (VDF + Signature)
            try self.engine.seal(block);

            // 5. Add to Chain
            try self.chain.add_block(block);

            std.debug.print("[Miner] Mined block {d}! Hash: {f}\n", .{ next_number, block.hash() });

            // Wait for next slot
            std.Thread.sleep(1000 * std.time.ns_per_ms);
        }
    }

    fn check_eligibility(self: *Miner, parent: *types.Block) !bool {
        _ = self;
        _ = parent;
        return true; // Stub: everyone mines
    }

    fn produce_block(self: *Miner, parent: *types.Block) !*types.Block {
        const next_number = parent.header.number + 1;
        const timestamp = @as(u64, @intCast(std.time.timestamp()));

        // 1. Fetch Txs from pool
        const txs = self.tx_pool.get_transactions(100); // returns []types.Transaction

        // 2. Construct Header
        const header = types.Header{
            .parent_hash = parent.hash(),
            .number = next_number,
            .time = timestamp,
            .verkle_root = parent.header.verkle_root, // Initial value
            .tx_hash = types.Hash.zero(),
            .coinbase = self.validator_addr,
            .extra_data = &[_]u8{},
            .gas_limit = parent.header.gas_limit,
            .gas_used = 0,
            .base_fee = parent.header.base_fee,
        };

        const block = try self.allocator.create(types.Block);
        block.* = types.Block{
            .header = header,
            .transactions = txs,
        };

        // 3. Execute Block (Updates Verkle Root)

        // Create slice of pointers for Executor interface
        var tx_ptrs = try self.allocator.alloc(*types.Transaction, txs.len);
        defer self.allocator.free(tx_ptrs);
        for (block.transactions, 0..) |*tx, i| {
            tx_ptrs[i] = tx;
        }

        const root = try self.executor.apply_block(self.state, &block.header, tx_ptrs);
        block.header.verkle_root = root; // Assign the returned Hash directly

        // Broadcast to P2P network
        if (self.p2p_server) |server| {
            const msg = p2p.types.NewBlockMsg{
                .block = block.*,
                .total_difficulty = 1, // Stub
                .hop_count = 0,
            };
            server.broadcast(p2p.types.MsgNewBlock, msg) catch |err| {
                std.debug.print("[Miner] Failed to broadcast block: {}\n", .{err});
            };
        }

        self.chain.set_head(block);

        // Remove executed transactions from pool
        self.tx_pool.remove_executed(txs);

        return block;
    }
};
